jmh-benchmarks/README.md

```diff
@@ -0,0 +1,196 @@
+###JMH-Benchmark module
+
+This module is used to contain benchmarks written using [JMH](http://openjdk.java.net/projects/code-tools/jmh/)
+from openjdk. Writing correct micro-benchmarks is Java (or another JVM language) is difficult
+and there are many non-obvious pitfalls (many due to compiler optimizations).  JMH is a framework for running
+and analyzing benchmarks (micro or macro) written in Java (or another JVM language). 
+
+For help in writing correct JMH tests, the best place to start is the [sample code](http://hg.openjdk.java.net/code-tools/jmh/file/tip/jmh-samples/src/main/java/org/openjdk/jmh/samples/) provided
+by the JMH project.
+
+Typically, JMH is expected to run as a separate project in Maven.  The jmh-benchmarks module uses
+the [gradle shadow jar](https://github.com/johnrengelman/shadow) plugin to emulate this behavior, by creating the required
+uber-jar file containing the benchmarking code and required JMH classes.  
+
+JMH is highly configurable and users are encouraged to look through the samples for suggestions
+on what options are available. A good tutorial for using JMH can be found [here](http://tutorials.jenkov.com/java-performance/jmh.html#return-value-from-benchmark-method)
+
+###Gradle Tasks / Running benchmarks in gradle
+
+If no benchmark mode is specified, the default is used which is throughput.
+
+*  jmh-benchmarks:shadowJar - creates the uber jar required to run the benchmarks.
+
+*  jmh-benchmarks:jmh - runs the clean, shadowJar tasks then runs all the benchmarks. 
+ 
+*  To run a specific benchmark (MemoryBytesCacheBenchmark.java for example): 
+        `./gradlew jmh-benchmarks:jmh -DjmhArgs=MemoryBytesCacheBenchmark` 
+
+*  To specify jmh specific flags, just add the desired flag and value (again no classname being specified runs all benchmarksHere).  Here
+     is an example for running the benchmark class (MemoryBytesCacheBenchmark.java) setting the "forks" property:
+          `./gradlew jmh-benchmarks:jmh -DjmhArgs=-f2,MemoryBytesCacheBenchmark`.  
+     The -f flag in JMH is the number of "forks" or test runs to do, with the default setting being 10.  Each fork or test run does 20 warmup iterations (default) and 20 benchmarking
+     iterations (again a default value).
+
+* By default all output goes to stdout.  If it is desired to capture the output in a file add -o&lt;filename&gt; to the jmhArgs property:
+   `./gradlew jmh-benchmarks:jmh -DjmhArgs=-f2,-oMemBytesBenchmark.txt,MemoryBytesCacheBenchmark`
+   
+
+ 
+### Running JMH outside of gradle
+The JMH benchmarks can be run outside of gradle as you would with any executable jar file:
+`java -jar <kafka-repo-dir>/jmh-benchmarks/build/libs/kafka-jmh-benchmarks-0.10.1.0-SNAPSHOT-all.jar -f2 MemoryBytesCacheBenchmark`
+
+### JMH Options
+
+To get all the jmh options you run jmh with the -h flag or just look below:
+
+```text
+java -jar jmh-benchmarks/build/libs/kafka-jmh-benchmarks-0.10.1.0-SNAPSHOT-all.jar -h
```

Comment:
It seems like this will get out of date pretty quickly. How about we include examples for the common cases and provide the command for people who want to get the full list?